internal/pkg/stream/cloudformation.go

```diff
@@ -0,0 +1,105 @@
+	"github.com/aws/aws-sdk-go/service/cloudformation"
+)
+
+// StackEventsDescriber is the CloudFormation interface needed to describe stack events.
+type StackEventsDescriber interface {
+	DescribeStackEvents(*cloudformation.DescribeStackEventsInput) (*cloudformation.DescribeStackEventsOutput, error)
+}
+
+// StackEvent is a CloudFormation stack event.
+type StackEvent struct {
+	LogicalResourceID string
+	ResourceType      string
+	ResourceStatus    string
+}
+
+// StackStreamer is a FetchNotifier for StackEvent events started by a change set.
+type StackStreamer struct {
+	Client                StackEventsDescriber // The CloudFormation client.
+	StackName             string               // The name of the stack to retrieve events from.
+	ChangeSetCreationTime time.Time            // The creation timestamp of the changeset being executed.
+
+	subscribers   []chan StackEvent
+	pastEventIDs  map[string]struct{}
+	eventsToFlush []StackEvent
+}
+
+// Subscribe registers the channels to receive notifications from the streamer.
+func (s *StackStreamer) Subscribe(channels ...chan StackEvent) {
+	s.subscribers = append(s.subscribers, channels...)
+}
+
+// Fetch retrieves and stores any new CloudFormation stack events since the ChangeSetCreationTime in chronological order.
+// If an error occurs from describe stack events, returns a wrapped error.
+func (s *StackStreamer) Fetch() (next time.Time, err error) {
+	if s.pastEventIDs == nil {
+		s.pastEventIDs = make(map[string]struct{})
+	}
+
+	var events []StackEvent
+	var nextToken *string
+	for {
+		out, err := s.Client.DescribeStackEvents(&cloudformation.DescribeStackEventsInput{
+			NextToken: nextToken,
+			StackName: aws.String(s.StackName),
+		})
+		if err != nil {
+			return next, fmt.Errorf("describe stack events %s: %w", s.StackName, err)
+		}
+
+		// Retrieve new events until we go past the ChangeSetCreationTime or we see an already seen event ID.
+		// This logic is taken from the AWS CDK:
+		// https://github.com/aws/aws-cdk/blob/43f3f09cc561fd32d651b2c327e877ad81c2ddb2/packages/aws-cdk/lib/api/util/cloudformation/stack-activity-monitor.ts#L230-L234
+		var finished bool
+		for _, event := range out.StackEvents {
+			if event.Timestamp.Before(s.ChangeSetCreationTime) {
+				finished = true
+				break
+			}
+			if _, seen := s.pastEventIDs[aws.StringValue(event.EventId)]; seen {
+				finished = true
+				break
+			}
+			events = append(events, StackEvent{
+				LogicalResourceID: aws.StringValue(event.LogicalResourceId),
+				ResourceType:      aws.StringValue(event.ResourceType),
+				ResourceStatus:    aws.StringValue(event.ResourceStatus),
+			})
+			s.pastEventIDs[aws.StringValue(event.EventId)] = struct{}{}
+		}
+		if finished || out.NextToken == nil {
+			break
+		}
+		nextToken = out.NextToken
+	}
+
+	// Store events to flush in chronological order.
+	reverse(events)
+	s.eventsToFlush = append(s.eventsToFlush, events...)
+	return time.Now().Add(3 * time.Second), nil
```

Comment:
Just wondering what this line is doing...?

Reply:
It's returning a date that's 3 seconds in the future indicating that `Fetch` should be called within 3 seconds.

I'll add a comment like this to make it clearer:
```suggestion
	return time.Now().Add(3 * time.Second) /* Next attempt should be in 3 seconds */, nil
```